{"nbformat":4,"nbformat_minor":0,"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.1"},"colab":{"name":"Копия блокнота \"02_Nonlinear_dimensionality_reduction.ipynb\"","provenance":[{"file_id":"https://github.com/KrishnaswamyLab/SingleCellWorkshop/blob/master/exercises/Dimensionality_Reduction/02_Nonlinear_dimensionality_reduction.ipynb","timestamp":1622227237091}]}},"cells":[{"cell_type":"markdown","metadata":{"id":"2VvEj0vjRVTk"},"source":["# Nonlinear dimensionality reduction\n","\n","## Goals\n","\n","* Visualize a single-cell dataset with t-SNE, UMAP and PHATE\n","* Understand how important parameter tuning is to visualization\n","* Understand how to compare the merits of different dimensionality reduction algorithms"]},{"cell_type":"code","metadata":{"id":"ZfKodsfLRVTm","executionInfo":{"status":"ok","timestamp":1622227183685,"user_tz":-300,"elapsed":8866,"user":{"displayName":"","photoUrl":"","userId":""}},"outputId":"069ac76e-9126-41f9-dee7-a3bdd782c25f","colab":{"base_uri":"https://localhost:8080/"}},"source":["!pip install scprep phate umap-learn"],"execution_count":1,"outputs":[{"output_type":"stream","text":["Collecting scprep\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/92/91/dfb35608ffae7eb0f45b565e702d3083d784988200ab6a967fd3f60d833e/scprep-1.1.0-py3-none-any.whl (104kB)\n","\r\u001b[K     |███▏                            | 10kB 13.5MB/s eta 0:00:01\r\u001b[K     |██████▎                         | 20kB 17.0MB/s eta 0:00:01\r\u001b[K     |█████████▍                      | 30kB 6.2MB/s eta 0:00:01\r\u001b[K     |████████████▌                   | 40kB 7.6MB/s eta 0:00:01\r\u001b[K     |███████████████▊                | 51kB 5.9MB/s eta 0:00:01\r\u001b[K     |██████████████████▉             | 61kB 5.4MB/s eta 0:00:01\r\u001b[K     |██████████████████████          | 71kB 5.9MB/s eta 0:00:01\r\u001b[K     |█████████████████████████       | 81kB 6.1MB/s eta 0:00:01\r\u001b[K     |████████████████████████████▎   | 92kB 4.9MB/s eta 0:00:01\r\u001b[K     |███████████████████████████████▍| 102kB 5.3MB/s eta 0:00:01\r\u001b[K     |████████████████████████████████| 112kB 5.3MB/s \n","\u001b[?25hCollecting phate\n","  Downloading https://files.pythonhosted.org/packages/0c/79/7f72c9d32bfd37a2d089fb4df8c50acb6e30813c33a7e9e982ef2f45c8d0/phate-1.0.7-py3-none-any.whl\n","Collecting umap-learn\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/75/69/85e7f950bb75792ad5d666d86c5f3e62eedbb942848e7e3126513af9999c/umap-learn-0.5.1.tar.gz (80kB)\n","\u001b[K     |████████████████████████████████| 81kB 8.6MB/s \n","\u001b[?25hRequirement already satisfied: decorator>=4.3.0 in /usr/local/lib/python3.7/dist-packages (from scprep) (4.4.2)\n","Requirement already satisfied: scipy>=0.18.1 in /usr/local/lib/python3.7/dist-packages (from scprep) (1.4.1)\n","Requirement already satisfied: pandas>=0.25 in /usr/local/lib/python3.7/dist-packages (from scprep) (1.1.5)\n","Requirement already satisfied: packaging in /usr/local/lib/python3.7/dist-packages (from scprep) (20.9)\n","Requirement already satisfied: scikit-learn>=0.19.1 in /usr/local/lib/python3.7/dist-packages (from scprep) (0.22.2.post1)\n","Requirement already satisfied: numpy>=1.12.0 in /usr/local/lib/python3.7/dist-packages (from scprep) (1.19.5)\n","Requirement already satisfied: matplotlib>=3.0 in /usr/local/lib/python3.7/dist-packages (from phate) (3.2.2)\n","Collecting graphtools>=1.3.1\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/dc/99/cc0d01f15c40656047caae88bbd3909ff535c116f2bba9e914423c8140e9/graphtools-1.5.2-py3-none-any.whl (45kB)\n","\u001b[K     |████████████████████████████████| 51kB 5.6MB/s \n","\u001b[?25hCollecting Deprecated\n","  Downloading https://files.pythonhosted.org/packages/fb/73/994edfcba74443146c84b91921fcc269374354118d4f452fb0c54c1cbb12/Deprecated-1.2.12-py2.py3-none-any.whl\n","Requirement already satisfied: future in /usr/local/lib/python3.7/dist-packages (from phate) (0.16.0)\n","Collecting tasklogger>=1.0\n","  Downloading https://files.pythonhosted.org/packages/5d/6b/cb2a724eff19829a0ada0217f403f54fca1e48c7de6fc3383e0a8b8fa121/tasklogger-1.0.0-py3-none-any.whl\n","Collecting s-gd2>=1.5\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/ab/46/aa500f0141a184f2b94f0b26aa01722b2d74bca50b1c4a02ee388aee2dc7/s_gd2-1.8-cp37-cp37m-manylinux2010_x86_64.whl (416kB)\n","\u001b[K     |████████████████████████████████| 419kB 38.9MB/s \n","\u001b[?25hRequirement already satisfied: numba>=0.49 in /usr/local/lib/python3.7/dist-packages (from umap-learn) (0.51.2)\n","Collecting pynndescent>=0.5\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/af/65/8189298dd3a05bbad716ee8e249764ff8800e365d8dc652ad2192ca01b4a/pynndescent-0.5.2.tar.gz (1.1MB)\n","\u001b[K     |████████████████████████████████| 1.2MB 37.5MB/s \n","\u001b[?25hRequirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas>=0.25->scprep) (2.8.1)\n","Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.7/dist-packages (from pandas>=0.25->scprep) (2018.9)\n","Requirement already satisfied: pyparsing>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging->scprep) (2.4.7)\n","Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.19.1->scprep) (1.0.1)\n","Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=3.0->phate) (0.10.0)\n","Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=3.0->phate) (1.3.1)\n","Collecting pygsp>=0.5.1\n","\u001b[?25l  Downloading https://files.pythonhosted.org/packages/d4/89/2f4aa73cccf12bec5179ac5d52a68b508120c838b7e5d456f5ea0c8beade/PyGSP-0.5.1-py2.py3-none-any.whl (1.8MB)\n","\u001b[K     |████████████████████████████████| 1.8MB 36.6MB/s \n","\u001b[?25hRequirement already satisfied: wrapt<2,>=1.10 in /usr/local/lib/python3.7/dist-packages (from Deprecated->phate) (1.12.1)\n","Requirement already satisfied: llvmlite<0.35,>=0.34.0.dev0 in /usr/local/lib/python3.7/dist-packages (from numba>=0.49->umap-learn) (0.34.0)\n","Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from numba>=0.49->umap-learn) (56.1.0)\n","Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.7.3->pandas>=0.25->scprep) (1.15.0)\n","Building wheels for collected packages: umap-learn, pynndescent\n","  Building wheel for umap-learn (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for umap-learn: filename=umap_learn-0.5.1-cp37-none-any.whl size=76569 sha256=3f1fda6da99041e46b4d8b07c173e8387b56325256e002473bdfa9a2d851ef2b\n","  Stored in directory: /root/.cache/pip/wheels/ad/df/d5/a3691296ff779f25cd1cf415a3af954b987fb53111e3392cf4\n","  Building wheel for pynndescent (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for pynndescent: filename=pynndescent-0.5.2-cp37-none-any.whl size=51351 sha256=c7600b92002154b35fc6cb1706e74b4e14412cb2b77279aaaaedb2e3c3987e24\n","  Stored in directory: /root/.cache/pip/wheels/ba/52/4e/4c28d04d144a28f89e2575fb63628df6e6d49b56c5ddd0c74e\n","Successfully built umap-learn pynndescent\n","Installing collected packages: scprep, pygsp, Deprecated, tasklogger, graphtools, s-gd2, phate, pynndescent, umap-learn\n","Successfully installed Deprecated-1.2.12 graphtools-1.5.2 phate-1.0.7 pygsp-0.5.1 pynndescent-0.5.2 s-gd2-1.8 scprep-1.1.0 tasklogger-1.0.0 umap-learn-0.5.1\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"V2Gl0J5bRVTn"},"source":["## 1. Loading the Retinal Bipolar dataset"]},{"cell_type":"code","metadata":{"id":"mjKuELWURVTn","executionInfo":{"status":"ok","timestamp":1622227184649,"user_tz":-300,"elapsed":968,"user":{"displayName":"","photoUrl":"","userId":""}}},"source":["import scprep\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","%matplotlib inline"],"execution_count":2,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"sps9CSO-RVTn"},"source":["Since we've done the preprocessing on this dataset before, we'll just download the preprocessed data from Google Drive.\n","\n","Alternatively, you may load your own data by replacing the Google Drive file ids with your own file ids.\n","\n","Note that if you do, you will likely not have annotated celltype labels yet. Replace all references to `metadata['CELLTYPE']` with an entry from `metadata`, or your favorite gene."]},{"cell_type":"code","metadata":{"id":"cynuqBiZRVTo","executionInfo":{"status":"ok","timestamp":1622227187028,"user_tz":-300,"elapsed":2382,"user":{"displayName":"","photoUrl":"","userId":""}}},"source":["scprep.io.download.download_google_drive(\"1GYqmGgv-QY6mRTJhOCE1sHWszRGMFpnf\", \"data.pickle.gz\")\n","scprep.io.download.download_google_drive(\"1q1N1s044FGWzYnQEoYMDJOjdWPm_Uone\", \"metadata.pickle.gz\")"],"execution_count":3,"outputs":[]},{"cell_type":"code","metadata":{"id":"ZJ-Dw1lkRVTo","executionInfo":{"status":"ok","timestamp":1622227190831,"user_tz":-300,"elapsed":3807,"user":{"displayName":"","photoUrl":"","userId":""}}},"source":["data = pd.read_pickle(\"data.pickle.gz\")\n","metadata = pd.read_pickle(\"metadata.pickle.gz\")"],"execution_count":4,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"w2HcA3kzRVTp"},"source":["## 2. t-SNE\n","\n","#### What is tSNE?\n","t-SNE is the most popular visualization method for single cell RNA-sequencing data. The method was first introduced by Laurens van der Maaten in 2008 in the aptly named article [\"Visualizing High-Dimensional Data Using t-SNE\"](http://jmlr.org/papers/v9/vandermaaten08a.html). The goal of t-SNE is to produce a two or three dimensional embedding of a dataset that exists in many dimensions such that the embedding can be used for visualization.\n","\n","By embedding, we're talking about projecting the data from high dimensions onto vectors in a smaller space.\n","\n","The way t-SNE does this is by minimizing the difference between neighborhood distances (i.e. distances from a cell to a set of close cells) in the original high dimensional space and the lower dimensional embedding space. t-SNE is an optimization problem where the algorithm iteratively learns a series of transformations such that each successive transformation better minimizes this difference between the high and low dimensional neighborhood distances. \n","\n","This approach preserves local structure in the data. Cells that are close in high dimensional space (i.e. have small Euclidean distances) will also be close in low dimensional space. However, it also means that global structure will not be preserved. This means that the distance between \"clusters\" in a t-SNE plot don't have any meaning.\n","\n","\n","#### How to use t-SNE effectively\n","\n","Unlike PCA, t-SNE has *hyperparameters* these are user-specified options that determine the output of t-SNE. Having hyperparameters isn't bad, but it is essential to understand what the hyperparameters are, what the effect of hyperpameter choices have on output, and how to select the best set of hyperparameters for a given research objective.\n","\n","In 2016, a group from Google Brain published great essay in Distill about [\"How to Use t-SNE Effectively\"](https://distill.pub/2016/misread-tsne/). In the article, they provide an interactive tool to explore the effect of various hyperparameters of t-SNE on various datasets.\n","\n","There are two main hyperparameters for t-SNE: **perplexity** and **learning rate** (sometimes called epsilon). Perplexity determines the \"neighborhood size\". Larger values of perplexity increase the number of points within the neighborhood. The reccomended range of t-SNE perplexity is roughly 5-50. Learning rate affects how quickly the algorithm \"stablilizes\". You probably don't need to change this, but should understand what it is."]},{"cell_type":"markdown","metadata":{"id":"7DQ_E-7qRVTq"},"source":["This dataset consists of many cell types, which were mostly identified as Amacrine cells, Muller Glia, Rod Bipolar cells, and many subtypes of Cone Bipolar cells in [Shekhar et. al, 2016](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5003425/). We can plot the data using t-SNE, as was done in the original paper."]},{"cell_type":"markdown","metadata":{"id":"D-mRB1VmRVTq"},"source":["#### Reducing dimensionality with PCA to speed up t-SNE\n","\n","t-SNE gets very slow with high-dimensional data. We can speed it up substantially by running PCA first to 100 dimensions."]},{"cell_type":"code","metadata":{"id":"uEdu-cVqRVTr"},"source":["data_pca = scprep.reduce.pca(data, n_components=100, method='dense')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"-KdE2QXMRVTr"},"source":["#### Subsampling to speed up t-SNE even more\n","\n","t-SNE is still slow even after PCA, so let's speed things up by using fewer points."]},{"cell_type":"code","metadata":{"id":"g0T9RtgERVTr"},"source":["data_pca_subsample, metadata_subsample = scprep.select.subsample(data_pca, metadata, n=3000)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"pWFI3yVNRVTr"},"source":["#### Running t-SNE\n","\n","tSNE is implemented in `scikit-learn`. t-SNE is a manifold learning algorithm and you can find the t-SNE operator at [`sklearn.manifold.TSNE`](https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html).\n","\n","We create a t-SNE operator and run it on data with the following syntax\n","\n","```python\n","import sklearn.manifold\n","tsne_op = sklearn.manifold.TSNE(n_components=2, perplexity=30)\n","data_tsne = tsne_op.fit_transform(data)\n","```"]},{"cell_type":"code","metadata":{"id":"oP7o8DqGRVTs"},"source":["import sklearn.manifold\n","tsne_op = sklearn.manifold.TSNE()\n","data_tsne = tsne_op.fit_transform(data_pca_subsample)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4qRHSwDPRVTs"},"source":["#### Plotting and interpreting t-SNE"]},{"cell_type":"code","metadata":{"id":"YLCVXx3wRVTs"},"source":["scprep.plot.scatter2d(data_tsne, c=metadata_subsample['CELLTYPE'],\n","                      figsize=(8,4), legend_anchor=(1,1),\n","                      ticks=False, label_prefix='t-SNE')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"vybQHXPWRVTs"},"source":["What do you notice? Is your favorite cell type nicely separated in this plot? How obvious is the distinction between the macro-level cell types of cone bipolar, rod bipolar, and glial cells?\n","\n","#### Exercise - run t-SNE with different `perplexity` parameters\n","\n","t-SNE's `perplexity` parameter describes the size of the neighborhood around each point. The authors recommend values between 5 and 100. Try a range of different values in and outside of this range and discuss the results with your group.\n","\n","*Note: be sure to use `data_pca_subsample`, as t-SNE can take a long time.*"]},{"cell_type":"code","metadata":{"id":"aQSPUYuwRVTt"},"source":["# ==============\n","# experiment with the perplexity parameter\n","tsne_op = sklearn.manifold.TSNE(perplexity=\n","data_tsne =\n","# =============="],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"b7MEOZ0WRVTt"},"source":["scprep.plot.scatter2d(data_tsne, c=metadata_subsample['CELLTYPE'],\n","                      figsize=(8,4), legend_anchor=(1,1), ticks=False, label_prefix='t-SNE')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gsAGRlDZRVTt"},"source":["## 3. UMAP\n","\n","Even though UMAP is not a part of scikit-learn, the syntax for UMAP is identical to t-SNE: `umap.UMAP().fit_transform`. UMAP is relatively fast, so you won't need to use the subsampled data. We also don't need to do PCA beforehand, but since we've already done it we may as well.\n","\n","Note, there's an article called [\"Understanding UMAP\" by Google Research](https://pair-code.github.io/understanding-umap/) that explores the effect of various parameters of UMAP in a similar way to the t-SNE artcile above."]},{"cell_type":"code","metadata":{"id":"Mcss1uRqRVTt"},"source":["import umap\n","umap_op = umap.UMAP()\n","data_umap = umap_op.fit_transform(data_pca)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"pYVqmwHBRVTt"},"source":["scprep.plot.scatter2d(data_umap, c=metadata['CELLTYPE'],\n","                      figsize=(8,4), legend_anchor=(1,1), ticks=False, label_prefix='UMAP')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"biwiQa3GRVTu"},"source":["What do you notice? Is your favorite cell type nicely separated in this plot? How obvious is the distinction between the macro-level cell types of cone bipolar, rod bipolar, and glial cells? How does this plot compare to t-SNE?"]},{"cell_type":"markdown","metadata":{"id":"8Z_YA79URVTu"},"source":["### Exercise - run UMAP with different `n_neighbors` and `min_dist` parameters\n","\n","UMAP's `n_neighbors` parameter describes the size of the neighborhood around each point. The `min_dist` parameter describes how tightly points can be packed together. The authors recommend values between 2 and 200 for `n_neighbors`, and between 0 and 0.99 for `min_dist`. Try a range of different values in and outside of these ranges and discuss the results with your group."]},{"cell_type":"code","metadata":{"id":"ViUrlK7MRVTu"},"source":["# ===============\n","# Choose different values for n_neighbors and min_dist, plotting with scprep\n","umap_op =\n","data_umap =\n","scprep.plot.scatter2d(\n","# ==============="],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"RaldkIZXRVTu"},"source":["## 4. PHATE\n","\n","### Exercise - perform PHATE and plot the results\n","\n","The syntax for PHATE is identical to UMAP and t-SNE: `phate.PHATE().fit_transform`. PHATE is relatively fast, so you won't need to use the subsampled data."]},{"cell_type":"code","metadata":{"id":"DmNn8e-ZRVTu"},"source":["import phate\n","phate_op = phate.PHATE()\n","data_phate = phate_op.fit_transform(data_pca)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"v7zgniDWRVTv"},"source":["scprep.plot.scatter2d(data_phate, c=metadata['CELLTYPE'],\n","                      figsize=(8,4), legend_anchor=(1,1), ticks=False, label_prefix='PHATE')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"f60wa0nfRVTv"},"source":["What do you notice? Is your favorite cell type nicely separated in this plot? How obvious is the distinction between the macro-level cell types of cone bipolar, rod bipolar, and glial cells? How does this plot compare to t-SNE and UMAP?"]},{"cell_type":"markdown","metadata":{"id":"o76rGHCGRVTv"},"source":["### Exercise - run PHATE with different `knn` and `t` parameters\n","\n","PHATE's `knn` parameter describes the size of the neighborhood around each point. The `t` parameter describes how much denoising is performed. We recommend values between 2 and 100 for `n_neighbors`, and between 2 and 150 for `t`. Try a range of different values in and outside of these ranges and discuss the results with your group."]},{"cell_type":"code","metadata":{"id":"1v8GaNiORVTv"},"source":["# ===============\n","# Choose different values for knn and y, plotting with scprep\n","phate_op =\n","data_phate =\n","scprep.plot.scatter2d(\n","# ==============="],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"74Zq2SGuRVTw"},"source":["### Discussion \n","\n","In groups, discuss the following questions:\n","1. In a dataset with clusters, how well does each method perform?\n","2. How might you determine which method is closest to the ground truth?\n","3. Which parameters are the most similar between methods?\n","4. Which method is the most / least sensitive to parameter selection?\n","5. If you run the same method with the same parameters multiple times, do you always get the same result?"]}]}